### ISSUE TITLE
Should not lint naming conventions on globally defined 

### ISSUE THREAD

**Issue by playpianolikewoah** [eslint/eslint#9005]
Currently we have a line something like 
```
__webpack_public_path__ = gon.webpack_public_path; // eslint-disable-line
```
We would like to define `__webpack_public_path__` as a global var in our `eslintrc` but eslint still complains of `camelCase`. That should not happen in my opinion. Global vars should be considered out of our control. I can understand checking for an override of a global, but not naming conventions of global vars.

**Comment by playpianolikewoah** [comment]
After defining the global in the `eslintrc`:

![image](https://user-images.githubusercontent.com/956301/28587529-7b56f2e6-7145-11e7-9c8a-11104c594611.png)

**Comment by eslintbot** [comment]
Hi @playpianolikewoah, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by mikegreiling** [comment]
I can fill in more details for you:

We are using eslint `v3.10.1`

We are using the AirBnB eslint ruleset which includes the following relevant rules: `camelcase` and `no-undef`.

This is the entire file in question:
```javascript
if (gon && gon.webpack_public_path) {
  __webpack_public_path__ = gon.webpack_public_path;
}
```

Our `.eslintrc` file contains:
```
  "globals": {
    "gon": false,
    "__webpack_public_path__": true,
  }
```

eslint complains on line 2 with `camelcase` and we were wondering if there was any way around this apart from disabling the line.

In theory, eslint could be smart enough to note that a global that is explicitly defined in `.eslintrc` shouldn't be subject to `camelcase`.

**Comment by not-an-aardvark** [comment]
Thanks for the info.

Hmm, I'm not sure about this -- it seems the camelcase rule is working as intended here. The goal of the rule is to encourage you to set variables that don't have underscores.

As an alternative to disabling all rules on the line, you could disable only the `camelcase` rule on that line using `// eslint-disable-line camelcase`.

**Comment by playpianolikewoah** [comment]
Thanks @not-an-aardvark. I think the main point is that in theory we should never have to disable eslint rules because they all are correct as long as we follow the directions. And in this situation although eslint is functioning I think eslint is in the wrong to correct naming conventions of something we declared as coming from an outside source. We can't always have control over outside sources. Of course we can disable the lines camelcase rule, no problem. But you wouldn't ask us to rename jQuery if it didn't fit the rules, because we can't rename jQuery. That's the proposal. 😄

**Comment by not-an-aardvark** [comment]
> I think the main point is that in theory we should never have to disable eslint rules because they all are correct as long as we follow the directions.

In general, this isn't really a goal. The rules are designed for the common use case, but there are exceptional cases where they don't apply, and it's appropriate to use disable comments for those.

> We can't always have control over outside sources.

I think the rule would intend for you do do something like this:

```js
const webpackPublicPath = window.__webpack_public_path__;

// use webpackPublicPath
```

Then you could use the `"properties": "never"` option to make the rule tolerate properties.

**Comment by mikegreiling** [comment]
In this case we're assigning something _to_ `__webpack_public_path__`, so I guess it would be more like `window.__webpack_public_path__ = foo;`, but I don't think this would actually work.  I'm fairly certain webpack converts `__webpack_public_path__` internally at compile time, so it never really exists as a property of `window`.

This is obviously a very obscure case, but to me it just seemed like if one defines a `global` variable that it should be assumed to be exempt from rules like `camelcase` by mere definition.  I can't think of a use case where the alternative would be preferred.  Why explicitly define a global that will always be flagged by your own rules when you try to use it?

**Comment by platinumazure** [comment]
I think it wouldn't hurt to enhance `camelcase` to take an array of strings that are Identifiers which should not be reported. (Not sure if @not-an-aardvark had a different idea in mind.)

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
